File: Leet-378. Kth Smallest Element in a Sorted Matrix/solution-binarysearch-nlogM.js
/**
 * @param {number[][]} matrix
 * @param {number} k
 * @return {number}
 */

// Time: O(nlogM), where M is the largest number in the matrix
var kthSmallest = function(matrix, k) {
    var m = matrix.length;
    var n = matrix[0].length;
    
    var left = matrix[0][0];
    var right = matrix[m - 1][n - 1];
    
    while (left <= right) {
        var mid = left + parseInt((right - left) / 2);
        var midExists = false;
        var count = 0;        
        
        var j = n - 1;
        for (var i = 0; i < m; i++) {
            // var j = n - 1;      // no need to reset j, since numbers to the bottom-right direction of matrix[i][j] is greater than matrix[i][j]
            while (j >= 0 && matrix[i][j] > mid) {
                j--;
            }

            if (matrix[i][j] === mid) {
                midExists = true;
            }            
            
            count += j + 1;
        }
        
        if (midExists && count === k ) {
            return mid;
        }
        
        if (count < k) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    return left;
};

File: Leet-378. Kth Smallest Element in a Sorted Matrix/solution-binarysearch-nlogM.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0 || k <= 0) {
            return 0;
        }
        
        int n = matrix.length;
        int left = matrix[0][0], right = matrix[n - 1][n - 1];
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int count = 0;
            int j = n - 1;
            boolean isExist = false;
            for (int i = 0; i < n; i++) {
                while (j >= 0 && matrix[i][j] > mid) {
                    j--;
                }
                
                if (j >= 0 && matrix[i][j] == mid) {
                    isExist = true;
                }
                
                count += j + 1;
            }
            
            if (isExist && count == k) {
                return mid;
            } else if (count < k) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        // in the last run, if there are other elements equal to the kth smallest element in the matrix,
        // left will equal to right. And in this case, count will > k, then, right will
        // be updated to be 1 less than left index.
        // However, the left index is the kth smallest element (due to the duplicate situation.)
        return left;
    }
}